src/stan/services/optimize/lbfgs.hpp

```diff
@@ -0,0 +1,206 @@
+       * @param[in] refresh how often to write output to message_writer
+       * @param[out] interrupt interrupt callback to be called every iteration
+       * @param[out] message_writer output for messages
+       * @param[out] init_writer Writer callback for unconstrained inits
+       * @param[out] parameter_writer output for parameter values
+       * @return stan::services::error_codes::OK if successful
+       */
+      template <class Model>
+      int lbfgs(Model& model,
+                stan::io::var_context& init,
+                unsigned int random_seed,
+                unsigned int chain,
+                double init_radius,
+                int history_size,
+                double init_alpha,
+                double tol_obj,
+                double tol_rel_obj,
+                double tol_grad,
+                double tol_rel_grad,
+                double tol_param,
+                int num_iterations,
+                bool save_iterations,
+                int refresh,
+                callbacks::interrupt& interrupt,
+                callbacks::writer& message_writer,
+                callbacks::writer& init_writer,
+                callbacks::writer& parameter_writer) {
+        boost::ecuyer1988 rng = stan::services::util::rng(random_seed, chain);
+
+        std::vector<int> disc_vector;
+        std::vector<double> cont_vector
+          = stan::services::util::initialize(model, init, rng, init_radius,
+                                             false,
+                                             message_writer, init_writer);
+
+        std::stringstream lbfgs_ss;
+        typedef stan::optimization::BFGSLineSearch
+          <Model, stan::optimization::LBFGSUpdate<> > Optimizer;
+        Optimizer lbfgs(model, cont_vector, disc_vector, &lbfgs_ss);
+        lbfgs.get_qnupdate().set_history_size(history_size);
+        lbfgs._ls_opts.alpha0 = init_alpha;
+        lbfgs._conv_opts.tolAbsF = tol_obj;
+        lbfgs._conv_opts.tolRelF = tol_rel_obj;
+        lbfgs._conv_opts.tolAbsGrad = tol_grad;
+        lbfgs._conv_opts.tolRelGrad = tol_rel_grad;
+        lbfgs._conv_opts.tolAbsX = tol_param;
+        lbfgs._conv_opts.maxIts = num_iterations;
+
+        double lp = lbfgs.logp();
+
+        std::stringstream msg;
+        msg << "Initial log joint probability = " << lp;
+        message_writer(msg.str());
+
+        std::vector<std::string> names;
+        names.push_back("lp__");
+        model.constrained_param_names(names, true, true);
+        parameter_writer(names);
+
+        if (save_iterations) {
+          std::vector<double> values;
+          msg.str("");
+          model.write_array(rng, cont_vector, disc_vector, values,
+                            true, true, &msg);
+          if (msg.str().length() > 0)
+            message_writer(msg.str());
+
+          values.insert(values.begin(), lp);
+          parameter_writer(values);
+        }
+        int ret = 0;
+
+        while (ret == 0) {
```

Comment:
There's a lot of repeated code pattern, but I don't want to duplicate my comments.  Can you just carry the earlier ones down, like making this a for loop?